File: modulo-govsocial/api/app/schemas/prontuario.py
import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


# ── Case files (prontuário) ───────────────────────────────────────
class CaseFileCreate(BaseModel):
    family_id: uuid.UUID
    unit_id: uuid.UUID
    service_type_code: str = Field(..., max_length=40)
    acolhida_data: Optional[date] = None
    acolhida_access_form_code: Optional[str] = None
    acolhida_motivo: Optional[str] = None
    acolhida_profissional_id: Optional[uuid.UUID] = None


class CaseFileUpdate(BaseModel):
    status: Optional[str] = None
    acolhida_data: Optional[date] = None
    acolhida_access_form_code: Optional[str] = None
    acolhida_motivo: Optional[str] = None
    acolhida_profissional_id: Optional[uuid.UUID] = None


class CaseFileListItem(BaseModel):
    id: uuid.UUID
    family_id: uuid.UUID
    unit_id: uuid.UUID
    service_type_code: str
    status: str
    acolhida_data: Optional[date]
    aberto_em: datetime
    created_at: datetime


class CaseFileOut(BaseModel):
    id: uuid.UUID
    family_id: uuid.UUID
    unit_id: uuid.UUID
    service_type_code: str
    status: str
    acolhida_data: Optional[date]
    acolhida_access_form_code: Optional[str]
    acolhida_motivo: Optional[str]
    acolhida_profissional_id: Optional[uuid.UUID]
    aberto_em: datetime
    created_at: datetime
    updated_at: datetime


# ── Attendances (atendimentos) ────────────────────────────────────
class AttendanceCreate(BaseModel):
    data_atendimento: datetime
    tipo: str = Field(..., max_length=30)
    evolution_text: Optional[str] = None
    sigiloso_reforcado: bool = False
    registrado_por_id: Optional[uuid.UUID] = None
    member_ids: list[uuid.UUID] = []
    professional_ids: list[uuid.UUID] = []


class AttendanceUpdate(BaseModel):
    data_atendimento: Optional[datetime] = None
    tipo: Optional[str] = None
    evolution_text: Optional[str] = None
    sigiloso_reforcado: Optional[bool] = None
    member_ids: Optional[list[uuid.UUID]] = None
    professional_ids: Optional[list[uuid.UUID]] = None


class AttendanceOut(BaseModel):
    id: uuid.UUID
    case_file_id: uuid.UUID
    unit_id: uuid.UUID
    service_type_code: str
    data_atendimento: datetime
    tipo: str
    sigiloso_reforcado: bool
    registrado_por_id: Optional[uuid.UUID]
    member_ids: list[uuid.UUID] = []
    professional_ids: list[uuid.UUID] = []
    # Evolução só é preenchida quando o usuário tem permissão de leitura.
    evolution_text: Optional[str] = None
    evolution_restrita: bool = False
    created_at: datetime
    updated_at: datetime


# ── Timeline / visão de rede ──────────────────────────────────────
class TimelineItem(BaseModel):
    attendance_id: uuid.UUID
    data_atendimento: datetime
    tipo: str
    service_type_code: str
    unit_id: uuid.UUID
    sigiloso_reforcado: bool
    pode_ler_evolucao: bool


class NetworkViewItem(BaseModel):
    """Visão de rede: outras unidades veem QUE houve atendimento, sem conteúdo."""

    unit_id: uuid.UUID
    unit_nome: Optional[str] = None
    service_type_code: str
    data_atendimento: datetime
    tipo: str


# ── Recepção ──────────────────────────────────────────────────────
class ReceptionCreate(BaseModel):
    unit_id: uuid.UUID
    person_id: Optional[uuid.UUID] = None
    family_id: Optional[uuid.UUID] = None
    nome_informado: Optional[str] = None
    motivo: Optional[str] = None
    senha: Optional[str] = None


class ReceptionUpdate(BaseModel):
    status: Optional[str] = None
    person_id: Optional[uuid.UUID] = None
    family_id: Optional[uuid.UUID] = None
    motivo: Optional[str] = None


class ReceptionOut(BaseModel):
    id: uuid.UUID
    unit_id: uuid.UUID
    data: datetime
    person_id: Optional[uuid.UUID]
    family_id: Optional[uuid.UUID]
    nome_informado: Optional[str]
    motivo: Optional[str]
    status: str
    senha: Optional[str]
    atendido_em: Optional[datetime]
    created_at: datetime


# ── Anexos ────────────────────────────────────────────────────────
class AttachmentOut(BaseModel):
    id: uuid.UUID
    case_file_id: uuid.UUID
    attendance_id: Optional[uuid.UUID]
    nome_arquivo: str
    tipo_documento: str
    content_type: Optional[str]
    tamanho_bytes: int
    versao: int
    enviado_por_id: Optional[uuid.UUID]
    created_at: datetime

File: modulo-govsocial/web/src/tipos/prontuario.ts
/**
 * DTOs do prontuário (case-files) e atendimentos — espelham o backend
 * (`app/schemas/prontuario.py`). Evolução técnica NUNCA trafega em listagens;
 * só no GET de um atendimento e apenas quando a política concede.
 */

export type TimelineItem = {
  attendance_id: string;
  data_atendimento: string;
  tipo: string;
  service_type_code: string;
  sigiloso_reforcado: boolean;
  pode_ler_evolucao: boolean;
  unit_id?: string;
  registrado_por_id?: string;
  registrado_por_nome?: string;
  situacao?: string;
  evolution_text?: string;
  member_ids?: string[];
};

export type CaseFileOut = {
  id: string;
  family_id: string;
  unit_id: string;
  service_type_code: string;
  status: string;
  acolhida_data?: string | null;
  acolhida_access_form_code?: string | null;
  acolhida_motivo?: string | null;
  acolhida_profissional_id?: string | null;
  aberto_em: string;
  created_at: string;
  updated_at: string;
};

export type AttendanceOut = {
  id: string;
  case_file_id: string;
  unit_id: string;
  service_type_code: string;
  data_atendimento: string;
  tipo: string;
  sigiloso_reforcado: boolean;
  registrado_por_id?: string | null;
  member_ids: string[];
  professional_ids: string[];
  /** Só é preenchida quando o usuário tem permissão de leitura. */
  evolution_text?: string | null;
  evolution_restrita: boolean;
  created_at: string;
  updated_at: string;
};

export type NetworkViewItem = {
  unit_id: string;
  unit_nome?: string | null;
  service_type_code: string;
  data_atendimento: string;
  tipo: string;
};

export type CaseFileListItem = {
  id: string;
  family_id: string;
  unit_id: string;
  service_type_code: string;
  status: string;
  acolhida_data?: string | null;
  aberto_em: string;
  created_at: string;
};

/** Motivos de desligamento do acompanhamento (alimentam o RMA do mês). */
export type MotivoDesligamento =
  | "OBJETIVOS_ALCANCADOS"
  | "MUDANCA_TERRITORIO"
  | "NAO_ADESAO"
  | "OBITO"
  | "ENCAMINHAMENTO_REDE"
  | "MEDIDA_ENCERRADA"
  | "TRANSFERENCIA_UNIDADE"
  | "OUTRO";
